internal/stacksapi/client.go

```diff
@@ -0,0 +1,261 @@
+package stacksapi
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/buildkite/agent-stack-k8s/v2/internal/version"
+	"github.com/buildkite/roko"
+)
+
+type Client struct {
+	baseURL    *url.URL
+	userAgent  string
+	authHeader string
+
+	logger      *slog.Logger
+	httpClient  *http.Client
+	retrierOpts []roko.RetrierOpt
+}
+
+func urlMustParse(u string) *url.URL {
+	parsed, err := url.Parse(u)
+	if err != nil {
+		panic(err)
+	}
+	return parsed
+}
+
+type ClientOpt func(*Client) error
+
+const (
+	DefaultBaseURL = "https://agent.buildkite.com/v3/"
+	DefaultQueue   = "_default"
+)
+
+var (
+	DefaultRetrierOptions = []roko.RetrierOpt{
+		roko.WithMaxAttempts(5),
+		roko.WithStrategy(roko.Constant(1 * time.Second)),
+		roko.WithJitterRange(500*time.Millisecond, 1*time.Second),
+	}
+
+	defaultUserAgent = "agent-stack-k8s/internal/stacksapi/" + version.Version()
+)
+
+func WithLogger(logger *slog.Logger) ClientOpt {
+	return func(c *Client) error {
+		c.logger = logger
+		return nil
+	}
+}
+
+// WithBaseURL sets the base URL for the API client, overriding [DefaultBaseURL]
+func WithBaseURL(baseURL *url.URL) ClientOpt {
+	return func(c *Client) error {
+		c.baseURL = baseURL
+		return nil
+	}
+}
+
+// PrependToUserAgent adds a prefix to the User-Agent header for the API client. Note that overriding of the user-agent
+// is not supported (though can probably be worked around).
+func PrependToUserAgent(prefix string) ClientOpt {
+	return func(c *Client) error {
+		c.userAgent = prefix + " " + c.userAgent
+		return nil
+	}
+}
+
+// WithHTTPClient sets the HTTP client for the API client.
+func WithHTTPClient(httpClient *http.Client) ClientOpt {
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// WithRetrierOptions sets the default retrier options for the API client. These can be overridden per-request by using the [WithRetrier] RequestOpt.
+func WithRetrierOptions(opts ...roko.RetrierOpt) ClientOpt {
+	return func(c *Client) error {
+		c.retrierOpts = opts
+		return nil
+	}
+}
```

Comment:
This might seem like a somewhat odd way to define a default retry strategy (as opposed to `WithRetrier(myRokoRetrier)`), but roko retriers have internal state, so it's not safe to share them across a client that may be executing multiple requests concurrently. happy to discuss better ways of doing this!